File: src/game/game_.py
from abc import ABCMeta, abstractmethod
from typing import Optional
from . import State, Action


class Game(metaclass=ABCMeta):
    def __init__(self, players):
        self.players = players

    DRAW = -1

    @abstractmethod
    def start(self) -> State:
        """Generate and return the initial game state."""
        pass

    @abstractmethod
    def legal_actions(self, state: State) -> list[Action]:
        """"Return all the possible actions from the given state."""
        pass

    @abstractmethod
    def next_state(self, state: State, move: Action) -> State:
        """Advance the given state and return it."""
        pass

    @abstractmethod
    def winner(self, state: State) -> int:
        """Return the winner of the game, None else."""
        pass

    def previous_player(self, current_player: int) -> int:
        """Return the previous turn player"""
        return (current_player - 1) % self.players

    @abstractmethod
    def get_action(self, from_state: State, to_state: State) -> Optional[Action]:
        """Return the action needed to go from one state to the following one"""
        pass

File: java/src/main/java/it/ai/game/Game.java
package it.ai.game;

import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;

public interface Game {

    /***
     * Generate and return the initial game state.
     * @return the initial game state
     */
    State start();

    /***
     * Return all the possible actions from the given state.
     */
    List<Action> getValidActions(State state);

    /***
     * Advance the given state and return it.
     */
    State nextState(State state, Action action);

    /***
     * Return the winner of the game, None else.
     */
    Optional<Integer> getWinner(State state);

    /***
     * Return the previous turn player
     */
    int previousPlayer(int player);

    /***
     * Return the next turn player
     */
    int nextPlayer(int player);

    Stream<Integer> getPlayers();

    /***
     * Return the action needed to go from one state to the following one.
     */
    Optional<Action> getAction(State fromState, State toState);


    /***
     * Return cells occupied by items of the given player
     */
    Stream<Coords> cellsOccupiedBy(State state, int player);
}
